### DISCUSSION TITLE
Thank you! and an issue...

### DISCUSSION THREAD

**Discussion by crazy880** [nickcoutsos/keymap-editor#167] [General]
Hi Nick and contributors.  This is a fantastic tool, thank you sincerely.  I have been editing for my new Wireless DAO CHOC BLE and have reached about 50 edits as I developed the layout.   **But the last thing i tried just destroyed two layers...!**

Explanation:
I have 4 layers - I had created a fourth for a 3L keyboard layout. i wanted to move it to layer 3, and move layer 3 (Bluetooth Layer) to layer 4.  I created layer 5, duplicated layer 3 keys and then deleted layer 3.  this caused the layer 4 (3L layout) to become layer 3 and layer 5 to become layer 4 - which seemed perfect! I checked a few items, found some orphaned combo mods, and fixed them.   Then I tried to save the config.

Immediately, layer 2 (number layer) layer 3 (former 3L layout) were deleted entirely.  Layer 4 which was the bluetooth layer after above changes, had now become layer 2.  I am now left with a default layer, and layer 1 for numbers/symbols and layer 3 for BT.  My number and 3L layers are both gone.

So, a couple of questions perhaps you could help with...?  

**1) how can I load the immediate prior config (before the catastrophe) saved in github back into the keymap-editor so i can continue using and evolving the config that I had arrived at before the attempted layer swap?**
2) how can I avoid this repeating, but still 'swap' my layer3 and layer4 seamlessly, for good?  
3) Is there a way to swap layers in the tool?

I look forward to your thoughts/help and thank you in advance for coming to the rescue. :)

Greg J

**Comment by nickcoutsos** [comment]
This falls out of scope of the app and would require a little Git management.

I'm assuming we're talking about this repository: https://github.com/crazy880/dao-zmk-config-new

I don't know if GitHub has anything in either the web interface or its desktop client, but if you have your repo cloned locally you can do

```
git reset --hard 3a4db157f5f5610e619ac42b3de78c174bad127c
```

and then push it back to GitHub with

```
git push origin main -f
```

Once you've got that the editor should load the old version.

As for the cause of the error, I see you have a layer name starting with a number. Devicetree syntax requires that a node name begins with a letter. The layer name field shows a validation error if this rule isn't followed but I'll need to make sure that it doesn't actually allow you to use the invalid name. Anyway, the syntax error causes problems with further parsing so the adding/deleting/moving of layer nodes can go off the rails unfortunately.

I'll try to bullet proof things a little more when I have a moment, but in the meantime steering clear of that should help.

**Reply by crazy880** [reply]
Thank you, Nick, immeasurably. I  was able to follow those instructions combined with my own search to revert a faulty commit using the web interface and got back to a healthy point.  Surely you can see that i am not a developer and just venturing into git and these other tools to accomplish a dream with this split keyboard... your quick reply is very much appreciated.

GJ

**Reply by crazy880** [reply]
BTW, i had also tried recovering the dao.keymap file and loading in from memory into the keymapper, which is where i went first in trying to figure out a way home... but i couldn't interpret a way to then save what i had uploaded to create a new commit.  is there a way to do that as well?

**Reply by nickcoutsos** [reply]
Hey, glad it worked out! Yeah, git doesn't really make it easy to dip your toes in for a single task. I like it ok as a developer but the main reason I integrate with GitHub is the automated builds. In hindsight, I guess another option was to copy the keymap source from the working commit and edit the current keymap file directly in GitHub to completely replace the contents.

I get the general idea of what you're saying in the second part but I'm not following the steps.

> but i couldn't interpret a way to then save what i had uploaded to create a new commit

Could you clarify the "what i had uploaded" part?

There's a lot of functionality that's currently exclusive to particular keymap source integrations (GitHub vs Clipboard vs FileSystem), but there's still a fair bit of potential overlap. I think somewhere down the road I'd like to flip things around a little so that you have more flexibility but that would need a lot of planning.
